### PR TITLE
Schema RPC to fetch table/view definition

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds a new RPC to the vttablet query service for VTGate to invoke to retrieve the updated view or table definition. 

## Related Issue(s)

- #11559 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 1]
```diff
@@ -1227,10 +1227,63 @@ func (qre *QueryExecutor) recordUserQuery(queryType string, duration int64) {
...
-func (qre *QueryExecutor) generateBindVarsForViewDDLInsert(createView *sqlparser.CreateView) map[string]*querypb.BindVariable {
+func generateBindVarsForViewDDLInsert(createView *sqlparser.CreateView) map[string]*querypb.BindVariable {
 	bindVars := make(map[string]*querypb.BindVariable)
 	bindVars["table_name"] = sqltypes.StringBindVariable(createView.ViewName.Name.String())
 	bindVars["view_definition"] = sqltypes.StringBindVariable(sqlparser.String(createView.Select))
 	bindVars["create_statement"] = sqltypes.StringBindVariable(sqlparser.String(createView))
 	return bindVars
 }
+
+func (qre *QueryExecutor) GetSchemaDef(tableType querypb.TableType, tableNames []string) (map[string]string, error) {
```
Comment (deepthi) [inline go/vt/vttablet/tabletserver/query_executor.go:1238]:
```suggestion
func (qre *QueryExecutor) GetSchemaDefinitions(tableType querypb.TableType, tableNames []string) (map[string]string, error) {
```

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 2]
```diff
@@ -1227,10 +1227,63 @@ func (qre *QueryExecutor) recordUserQuery(queryType string, duration int64) {
...
+	switch tableType {
+	case querypb.TableType_VIEWS:
+		return qre.getViewDef(tableNames)
+	case querypb.TableType_TABLES:
+		return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid table type %v", tableType)
+	}
+	return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid table type %v", tableType)
+}
+
+func (qre *QueryExecutor) getViewDef(tableNames []string) (map[string]string, error) {
```
Comment (deepthi) [inline go/vt/vttablet/tabletserver/query_executor.go:1248]:
```suggestion
func (qre *QueryExecutor) getViewDefinitions(tableNames []string) (map[string]string, error) {
```

Comment (deepthi) [reply go/vt/vttablet/tabletserver/query_executor.go:1248]: shouldn't it be `viewNames` instead of `tableNames`?

File: proto/query.proto [hunk 3]
```diff
@@ -973,3 +973,27 @@ message TransactionMetadata {
   int64 time_created = 3;
   repeated Target participants = 4;
 }
+
+
+// TableType represents the type of table requested.
+enum TableType {
+  VIEWS = 0;
+  TABLES = 1;
+  ALL_TABLES_VIEWS = 2;
```
Comment (deepthi) [inline proto/query.proto:982]: can be just `ALL`?
that way if we add any other type of objects in the future, the name doesn't become obsolete.

**Review by deepthi — COMMENTED** [review]
This comment in schema.go needs to be updated
```
// This file provides a few utility variables and methods, mostly for tests.
// The assumptions made about the types of fields and data returned
// by MySQl are validated in schema_test.go. This way all tests
// can use these variables and methods to simulate a MySQL server
// (using fakesqldb/ package for instance) and still be guaranteed correct
// data.
```
We are now using those queries in production, not just tests.

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/query_executor.go:1248]: everything is a table only the type is different. `VIEW / BASE TABLE` i.e. how MySQL differentiates them.

File: go/vt/vtgate/schema/tracker.go [hunk 4]
```diff
@@ -300,21 +300,11 @@ func (t *Tracker) updateTables(keyspace string, res *sqltypes.Result) {
...
-	res, err := th.Conn.Execute(t.ctx, th.Target, mysql.FetchUpdatedViews, bv, 0, 0, nil)
+	res, err := th.Conn.GetSchema(t.ctx, th.Target, querypb.SchemaTableType_VIEWS, viewsUpdated)
 	if err != nil {
 		t.tracked[th.Target.Keyspace].setLoaded(false)
 		// TODO: optimize for the views that got errored out.
 		log.Warningf("error fetching new views definition for %v", viewsUpdated, err)
-		code := vterrors.Code(err)
-		if code == vtrpcpb.Code_UNAUTHENTICATED || code == vtrpcpb.Code_PERMISSION_DENIED {
-			log.Warning(aclErrorMessageLog)
-		}
```
Comment (frouioui) [inline go/vt/vtgate/schema/tracker.go:317]: What's the reason for removing this?

File: go/vt/vttablet/endtoend/views_test.go [hunk 5]
```diff
@@ -145,7 +148,30 @@ func TestDropViewDDL(t *testing.T) {
...
+	viewSchemaDef, err := client.GetSchema(querypb.SchemaTableType_VIEWS)
+	require.NoError(t, err)
+	require.Zero(t, len(viewSchemaDef))
+
+	client.UpdateContext(callerid.NewContext(
+		context.Background(),
+		&vtrpcpb.CallerID{},
+		&querypb.VTGateCallerID{Username: "dev"}))
+
+	defer client.Execute(qDelAllRows, nil)
```
Comment (frouioui) [inline go/vt/vttablet/endtoend/views_test.go:169]: nit: shoudn't we do this after `client.Execute` and before the `require.NoError`

File: go/vt/vtgate/schema/tracker.go (hunk 4 above)
Comment (harshit-gangal) [reply go/vt/vtgate/schema/tracker.go:317]: GetSchema does not go through ACL check, so tracker will not see such error message.

Comment (frouioui) [reply go/vt/vtgate/schema/tracker.go:317]: Got it

**Review by frouioui — APPROVED** [review]

**Review by GuptaManan100 — APPROVED** [review]

**Pull request merged**